_test/check-exercises.sh

```diff
@@ -1,73 +1,56 @@
+# $ BENCHMARK=1 rustup run nightly _test/check-exercises.sh
 if [ -n "$BENCHMARK" ]; then
     files=exercises/*/benches
+    # cargo bench --release is an invalid command
+    release=""
 else
     files=exercises/*/tests
+    release="--release"
 fi
 
-tmp=${TMPDIR:-/tmp/}
-mkdir "${tmp}exercises"
-
-exitcode=0
-
 # An exercise worth testing is defined here as any top level directory with
 # a 'tests' directory
 for exercise in $files; do
-    # This assumes that exercises are only one directory deep
-    # and that the primary module is named the same as the directory
-    directory=$(dirname "${exercise}");
-
-    workdir=$(mktemp -d "${tmp}${directory}.XXXXXXXXXX")
-
-    cp -R -T $directory $workdir
-
-    # Run in subshell to change workdir without affecting current workdir.
-    (
-        cd $workdir
-
-        cp example.rs src/lib.rs
-
-        # Overwrite empty Cargo.toml if an example specific file exists
-        if [ -f Cargo-example.toml ]; then
-            cp Cargo-example.toml Cargo.toml
-        fi
-
-        # Forcibly strip all "ignore" statements from the testing files
-        for test in tests/*.rs; do
-            sed -i '/\[ignore\]/d' $test
-        done
-
-        # Run benchmarks instead of tests when enabled.
-        if [ -n "$BENCHMARK" ]; then
-            cargo bench
-        elif [ -n "$DENYWARNINGS" ]; then
-            sed -i -e '1i #![deny(warnings)]' src/lib.rs
-
-            # No-run mode so we see no test output.
-            # Quiet mode so we see no compile output
-            # (such as "Compiling"/"Downloading").
-            # Compiler errors will still be shown though.
-            # Both flags are necessary to keep things quiet.
-            cargo test --quiet --no-run
-        else
-            # Run the test and get the status
-            cargo test
-        fi
-    )
-
-    status=$?
-
-    if [ $status -ne 0 ]
-    then
-        exitcode=1
-    fi
+   # This assumes that exercises are only one directory deep
+   # and that the primary module is named the same as the directory
+   directory=$(dirname "${exercise}");
+
+   if [ -n "$DENYWARNINGS" ]; then
+      sed -i -e '1i #![deny(warnings)]' "$directory/src/lib.rs"
+      # No-run mode so we see no test output.
+      # Quiet mode so we see no compile output
+      # (such as "Compiling"/"Downloading").
+      # Compiler errors will still be shown though.
+      # Both flags are necessary to keep things quiet.
+      ./bin/test-exercise $directory --quiet --no-run
+   else
+      # Run the test and get the status
+      # We use release mode here because, while it somewhat increases
+      # the compile time for all exercises, it substantially improves
+      # the runtime for certain exercises such as alphametics.
+      # Overall this should be an improvement.
+      ./bin/test-exercise $directory $release
+   fi
+   # as the test-exercise command is the last one run in all cases,
+   # its exit code is preserved automatically
```

Comment:
true statement, but we do not have `-e` in DENYWARNINGS mode, therefore the exit code is the exit code of that of the last exercise, is it not? Thus, we would only get a nonzero exit if the very last exercise has any warnings, and all other exercises' warnings are ignored, I think?

Reply:
I see. Just tested, and you're correct. I'll update that.